### PR TITLE
[WIP] feat: Add "deno jupyter" subcommand

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/13016

TODO:
- [x] Add support for `DEBUG` env variable
- [ ] Make stdio out less flaky (don't use channel for all messages)
- [ ] Investigate `BrokenPipe` errors we're getting and how we could improve that situation (https://github.com/zeromq/zmq.rs/issues/153)

### CODE REVIEW

File: cli/tools/jupyter.rs [hunk 1]
```diff
@@ -0,0 +1,893 @@
...
+  metadata: Value,
+  content: Value,
+  session_id: String,
+}
+
+impl Message {
+  fn new() -> Self {
+    todo!()
+  }
+
```
Comment (bnoordhuis) [inline cli/tools/jupyter.rs:340]:
```suggestion
```

File: cli/tools/jupyter.rs [hunk 2]
```diff
@@ -0,0 +1,893 @@
...
+use deno_core::serde_json::json;
+use deno_core::serde_json::Value;
+use ring::hmac;
+use std::env::current_exe;
+use tempfile::TempDir;
+use tokio::join;
+use zeromq::prelude::*;
+use zeromq::ZmqMessage;
+
+const DELIMETER: &str = "<IDS|MSG>";
```
Comment (bnoordhuis) [inline cli/tools/jupyter.rs:24]:
```suggestion
const DELIMITER: &str = "<IDS|MSG>";
```

File: cli/tools/jupyter.rs [hunk 3]
```diff
@@ -0,0 +1,893 @@
...
+    } else {
+      serde_json::to_string(&json!({})).unwrap()
+    };
+    let metadata = serde_json::to_string(&self.metadata).unwrap();
+    let content = serde_json::to_string(&self.content).unwrap();
+
+    let hmac =
+      hmac_sign(hmac_key, &header, &parent_header, &metadata, &content);
+
+    let mut zmq_msg = ZmqMessage::from(DELIMETER);
```
Comment (bnoordhuis) [inline cli/tools/jupyter.rs:354]:
```suggestion
    let mut zmq_msg = ZmqMessage::from(DELIMITER);
```

File: cli/tools/jupyter.rs [hunk 4]
```diff
@@ -0,0 +1,893 @@
...
+
+  let mut sock = zeromq::RepSocket::new(); // TODO(apowers313) exact same as dealer, refactor
+  sock.monitor();
+  sock.bind(conn_str).await?;
+
+  loop {
+    let msg = sock.recv().await?;
+    dbg!(&msg);
+    println!("{} got packet!", name);
+  }
```
Comment (bnoordhuis) [inline cli/tools/jupyter.rs:494]: Double-checking: this recv-and-drop is intentional?

Aside: `dbg!()` _and_ `println!()` is kind of excessive.

File: cli/tools/jupyter.rs [hunk 5]
```diff
@@ -0,0 +1,893 @@
...
+  let _metadata = data.get(4);
+  let _content = data.get(5);
+
+  println!("header:");
+  dbg!(header);
+  let header_str = std::str::from_utf8(header).unwrap();
+  let header_value: MessageHeader = serde_json::from_str(header_str).unwrap();
+  println!("header_value");
+  dbg!(&header_value);
+  // validate_header(&header_value)?;
```
Comment (bnoordhuis) [inline cli/tools/jupyter.rs:515]: I assume this still needs to be fleshed out further?

File: cli/tools/jupyter.rs [hunk 6]
```diff
@@ -0,0 +1,893 @@
...
+  parent_header: &[u8],
+  metadata: &[u8],
+  content: &[u8],
+) -> Result<(), ring::error::Unspecified> {
+  let mut msg = Vec::<u8>::new();
+  msg.extend(header);
+  msg.extend(parent_header);
+  msg.extend(metadata);
+  msg.extend(content);
+  let sig = HEXLOWER.decode(expected_signature).unwrap();
```
Comment (bnoordhuis) [inline cli/tools/jupyter.rs:550]:
```suggestion
  let sig = HEXLOWER.decode(expected_signature)?;
```
unwrap() seems a little unsafe.

File: cli/tools/jupyter.rs [hunk 7]
```diff
@@ -0,0 +1,893 @@
...
+        }
+      }
+      Err(err) => {
+        eprintln!("Failed to install kernelspec: {}", err);
+      }
+    }
+  }
+
+  let _ = std::fs::remove_dir(temp_dir);
+  println!("Deno kernelspec installed successfully.");
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/tools/jupyter.rs:69]: Still wip, I'll fix this before landing

File: cli/tools/jupyter.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/tools/jupyter.rs:494]: Still WIP

File: cli/tools/jupyter.rs [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+#[derive(Debug)]
+struct Message {
+  is_reply: bool,
+  r#type: String,
+  header: MessageHeader,
+  parent_header: Option<MessageHeader>,
+  metadata: Value,
+  content: Value,
+  session_id: String,
+}
```
Comment (bartlomieju) [inline cli/tools/jupyter.rs:335]: Feel like this should be split into `Message` and `ReplyMessage`. AFAIK `parent_header` is only present in response messages and for messages received from notebook are always empty dict (`{}`). We could also remove `session_id` and `is_reply` for `Message`. @apowers313 WDYT?

Comment (bartlomieju) [reply cli/tools/jupyter.rs:335]: `r#type` is also fishy - it's the same value as `header.msg_type`...

Comment (apowers313) [reply cli/tools/jupyter.rs:335]: Refactoring into `RequestMessage` and `ReplyMessage` and creating `From` traits for more more specific types.

**Comment by apowers313** [comment]
Trying to do this all in a single file is becoming cumbersome. Can we create a `cli/jupyter` folder and start creating multiple files there?

**Comment by bartlomieju** [comment]
> Trying to do this all in a single file is becoming cumbersome. Can we create a `cli/jupyter` folder and start creating multiple files there?

Yes, that sounds good to me

File: Cargo.lock [hunk 9]
```diff
@@ -4974,3 +5145,29 @@ dependencies = [
...
+name = "zeromq"
+version = "0.3.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "66b9ebf6a24cc77c2c7053f8f0df07e6261abc5166c060d874efc828080c7316"
+dependencies = [
+ "async-trait",
+ "asynchronous-codec",
+ "bytes",
+ "crossbeam",
+ "dashmap 3.11.10",
```
Comment (ry) [inline Cargo.lock:5159]: Multiple versions of this special hashmap optimized for parallel access in our single threaded program. Sigh.

File: cli/tools/jupyter/install.rs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+pub fn install() -> Result<(), AnyError> {
+  let temp_dir = TempDir::new().unwrap();
+  let kernel_json_path = temp_dir.path().join("kernel.json");
+
+  // TODO(bartlomieju): add remaining fields as per
+  // https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs
+  // FIXME(bartlomieju): replace `current_exe`
+  let json_data = json!({
+      "argv": [current_exe().unwrap().to_string_lossy(), "jupyter", "--conn", "{connection_file}"],
+      "display_name": "Deno (Rust)",
```
Comment (bartlomieju) [inline cli/tools/jupyter/install.rs:18]:
```suggestion
      "display_name": "Deno",
```

File: core/ops.rs [hunk 11]
```diff
@@ -201,7 +201,16 @@ impl OpTable {
...
+    assert!(prev.is_none(), "Op '{}' was already registered", name);
+    op_id
+  }
+
+  pub fn overwrite_op<F>(&mut self, name: &str, op_fn: F) -> OpId
+  where
+    F: Fn(Rc<RefCell<OpState>>, OpPayload) -> Op + 'static,
+  {
+    let (op_id, prev) = self.0.insert_full(name.to_owned(), Rc::new(op_fn));
+    assert!(prev.is_some(), "Op '{}' wasn't already registered", name);
```
Comment (bartlomieju) [inline core/ops.rs:213]: Upstream in separate PR

File: core/runtime.rs [hunk 12]
```diff
@@ -637,6 +637,18 @@ impl JsRuntime {
...
+    F: Fn(Rc<RefCell<OpState>>, OpPayload) -> Op + 'static,
+  {
+    Self::state(self.v8_isolate())
+      .borrow_mut()
+      .op_state
+      .borrow_mut()
+      .op_table
+      .overwrite_op(name, op_fn)
+  }
+
```
Comment (bartlomieju) [inline core/runtime.rs:651]: Upstream in separate PR

**Comment by bartlomieju** [comment]
Note to self: try https://github.com/cetra3/tmq

**Pull request closed without merging**